### PR TITLE
Harshp8l master

### PR DESCRIPTION
**Pull request by nehagjain15**
# Description

Please include a summary of the change and which issue is fixed. Please also include relevant motivation and context. List any dependencies that are required for this change.

## Related PR and Issues
Fixes # (issue)

## Impacted Areas in Application
List general components of the application that this PR will affect:

## Type of change

Please delete options that are not relevant.

- [ ] Bug fix (non-breaking change which fixes an issue)
- [ ] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to not work as expected)
- [ ] This change requires a documentation update

## How Has This Been Tested?

Please describe the tests that you ran to verify your changes. Provide instructions so we can reproduce. Please also list any relevant details for your test configuration

- [ ] Test A
- [ ] Test B

**Test Configuration**:
* Environment Variables for Integration test:
* Hardware Requirements in ESXi:
* Toolchain:
* SDK:

# Checklist:

- [ ] My code follows the standard ruby style guide
- [ ] I have performed a self-review of my own code
- [ ] I have commented my code, particularly in hard-to-understand areas
- [ ] I have added tests that prove my fix is effective or that my feature works
- [ ] New and existing unit tests pass locally with my changes

### CODE REVIEW

File: src/vsphere_cpi/spec/integration/gpu_parallel_spec.rb [hunk 1]
```diff
@@ -0,0 +1,310 @@
...
+    end
+    context 'when ephemeral datastore is accessible from GPU host' do
+      let(:vm_type_single_cluster_acc_ds) do
+        vm_type_single_cluster.merge(
+            'datastores' => [@gpu_host_datastore_one]
+        )
+      end
+
+      # parallel vm calls
+      context 'when vm_type has two clusters specifies 1 GPU and we create 6 vms in parallel' do
```
Comment (nehagjain15) [inline src/vsphere_cpi/spec/integration/gpu_parallel_spec.rb]: Looks like context are not arranged properly, the outermost context says ' when vm_type specifies single cluster and this says when it has 2 clusters'. This context should not be nested inside the outermost context

Comment (nehagjain15) [reply src/vsphere_cpi/spec/integration/gpu_parallel_spec.rb]: going through all the context and let is confusing, this needs to be cleaned up

File: src/vsphere_cpi/spec/integration/gpu_parallel_spec.rb [hunk 2]
```diff
@@ -0,0 +1,310 @@
...
+  before (:all) do
+    @datacenter_name = fetch_and_verify_datacenter('BOSH_VSPHERE_CPI_DATACENTER')
+    @cluster_name = fetch_and_verify_cluster('BOSH_VSPHERE_CPI_CLUSTER')
+    @second_cluster_name = fetch_and_verify_cluster('BOSH_VSPHERE_CPI_SECOND_CLUSTER')
+    @gpu_host_datastore_one = fetch_property('BOSH_VSPHERE_CPI_SHARED_DATASTORE')
+    @host_1 = '10.114.22.249' #4 GRID K1 GPUs
+    @host_2 = '10.114.22.29' #2 GRID K2 GPUs
+  end
+
+  let(:base_vm_type) do
```
Comment (nehagjain15) [inline src/vsphere_cpi/spec/integration/gpu_parallel_spec.rb]: the base_vm_type can have properties from vm_type_single_cluster as well as that is what is being used in all tests below

File: src/vsphere_cpi/spec/integration/gpu_parallel_spec.rb [hunk 3]
```diff
@@ -0,0 +1,310 @@
...
+            begin
+              vm_id_1 = cpi.create_vm(
+                  'agent-007',
+                  @stemcell_id,
+                  vm_type_single_cluster_acc_ds_1_gpu,
+                  get_network_spec,
+                  [],
+                  {}
+              )
+              expect(vm_id_1).to_not be_nil
```
Comment (nehagjain15) [inline src/vsphere_cpi/spec/integration/gpu_parallel_spec.rb]: this check is not needed

File: src/vsphere_cpi/spec/integration/gpu_parallel_spec.rb [hunk 4]
```diff
@@ -0,0 +1,310 @@
...
+
+              vm = cpi.vm_provider.find(vm_id_1)
+              expect(vm).to_not be_nil
```
Comment (nehagjain15) [inline src/vsphere_cpi/spec/integration/gpu_parallel_spec.rb]: this check is also not needed as assignment will raise error if vm is not found

Comment (nehagjain15) [reply src/vsphere_cpi/spec/integration/gpu_parallel_spec.rb]: same applies to all vm checks in this test

File: src/vsphere_cpi/spec/unit/cloud/vsphere/resources/cluster_spec.rb [hunk 5]
```diff
@@ -27,9 +28,10 @@ module VSphereCloud::Resources
     let(:properties) do
       {
         :obj => cluster_mob,
-        'host' => cluster_hosts,
+        #'host' => cluster_hosts,
```
Comment (nehagjain15) [inline src/vsphere_cpi/spec/unit/cloud/vsphere/resources/cluster_spec.rb]: Remove this line

**Comment by EleanorRigby** [comment]
Open Later for gpu work if needed.

**Pull request closed without merging**